clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -459,19 +459,14 @@ boolean joinGroupIfNeeded(final Timer timer) {
             }
 
             if (future.succeeded()) {
-                Generation generationSnapshot;
-                MemberState stateSnapshot;
-
                 // Generation data maybe concurrently cleared by Heartbeat thread.
                 // Can't use synchronized for {@code onJoinComplete}, because it can be long enough
                 // and shouldn't block heartbeat thread.
                 // See {@link PlaintextConsumerTest#testMaxPollIntervalMsDelayInAssignment}
-                synchronized (AbstractCoordinator.this) {
-                    generationSnapshot = this.generation;
-                    stateSnapshot = this.state;
-                }
+                Generation generationSnapshot;
+                MemberState stateSnapshot = null;
 
-                if (!hasGenerationReset(generationSnapshot) && stateSnapshot == MemberState.STABLE) {
+                if (!hasGenerationReset(generationSnapshot = this.generation) && (stateSnapshot = this.state) == MemberState.STABLE) {
```

Comment:
I think, there is still a possibility of race condition here. `generation` gets updated by the heartbeat thread in a synchronized block, but there's no guarantee that what we see here might be the updated value because not we are not even holding any locks.